clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1437,20 +1500,23 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
                                                        : CalculatedName);
 
   SyclKernelFieldChecker checker(*this);
-  SyclKernelDeclCreator kernel_decl(*this, checker, KernelName,
-                                    KernelLambda->getLocation(),
-                                    KernelCallerFunc->isInlined());
-  SyclKernelBodyCreator kernel_body(*this, kernel_decl, KernelLambda,
-                                    KernelCallerFunc);
-  SyclKernelIntHeaderCreator int_header(
-      *this, getSyclIntegrationHeader(), KernelLambda,
-      calculateKernelNameType(Context, KernelCallerFunc), KernelName,
-      StableName);
-
-  ConstructingOpenCLKernel = true;
-  VisitRecordFields(KernelLambda->fields(), checker, kernel_decl, kernel_body,
-                    int_header);
-  ConstructingOpenCLKernel = false;
+  VisitRecordFields(KernelLambda->fields(), checker);
+  if (checker.isValid()) {
```

Comment:
I don't know if you saw it last time, but the intent for this case is instead to have the actual 'visit' methods return a bool, which determines whether each field/entry should be continued on each.  Without C++17 it takes a bit of template work to do a && fold (instead of the init list trick), but we should do that instead.